### PR TITLE
Update densenet_efficient.py

### PR DESCRIPTION
**Pull request by wandering007**
simplify the BN statistics restoring; forward function can be more efficient by creating volatile Variables

### CODE REVIEW

File: models/densenet_efficient.py [hunk 1]
```diff
@@ -300,16 +300,11 @@ def __init__(self, shared_allocation_1, shared_allocation_2,
...
-        self.curr_running_var = self.running_var.new(self.running_var.size())
-
     def forward(self, bn_weight, bn_bias, *inputs):
-        # Save the current BN statistics for later
-        self.prev_running_mean.copy_(self.running_mean)
-        self.prev_running_var.copy_(self.running_var)
+        if self.training:
+            # Save the current BN statistics for later
+            prev_running_mean = self.running_mean.clone()
+            prev_running_var = self.running_var.clone()
```
Comment (gpleiss) [inline models/densenet_efficient.py]: Rather than calling `.clone` here, it'll be faster if we have a buffer that we copy into. So i'd prefer to keep it the way it currently is.

File: models/densenet_efficient.py [hunk 2]
```diff
@@ -319,15 +314,15 @@ def forward(self, bn_weight, bn_bias, *inputs):
...
         relu_output = type(inputs[0])(storage).resize_(size)
 
         # Create variable, using existing storage
         torch.cat(inputs, dim=1, out=bn_input_var.data)
 
         # Do batch norm
-        bn_weight_var = Variable(bn_weight, requires_grad=False)
-        bn_bias_var = Variable(bn_bias, requires_grad=False)
+        bn_weight_var = Variable(bn_weight, volatile=True)
+        bn_bias_var = Variable(bn_bias, volatile=True)
```
Comment (gpleiss) [inline models/densenet_efficient.py]: Good catch with the `volatile`s!

File: models/densenet_efficient.py [hunk 3]
```diff
@@ -300,16 +300,11 @@ def __init__(self, shared_allocation_1, shared_allocation_2,
         self.momentum = momentum
         self.eps = eps
 
-        # Buffers to store old versions of bn statistics
-        self.prev_running_mean = self.running_mean.new(self.running_mean.size())
-        self.prev_running_var = self.running_var.new(self.running_var.size())
-        self.curr_running_mean = self.running_mean.new(self.running_mean.size())
-        self.curr_running_var = self.running_var.new(self.running_var.size())
```
Comment (gpleiss) [inline models/densenet_efficient.py]: You're right that `curr_running_mean` and `curr_running_var` can be deleted, but I'd prefer to keep `prev_running_mean` rather than using `.clone()`. (See below).

**Comment by wandering007** [comment]
@gpleiss Thanks, I've made the changes as suggested.

**Comment by gpleiss** [comment]
Thanks @wandering007!

**Comment by panjunwen** [comment]
How could I save the mean and var ? 
When recreate a new network and load weights from disk, the mean and var should be reload correctly.

**Comment by gpleiss** [comment]
They should already be included as part of the state dict. Are they not?

**Comment by panjunwen** [comment]
Thanks for your reply. And moving averages is done by F.batch_norm() as well?
I use your code for medical image segmentation. I found that when using eval mode, the network's parformance is really bad.  I must forward several data in training mode and it could  be better.
I alse found that in keras the default value of momentum of bacth norm is 0.99 but in pytorch only 0.1.

**Comment by gpleiss** [comment]
Moving averages are handled by F.batch_norm() as well. You can check out http://pytorch.org/docs/stable/nn.html#batchnorm2d for more info.

> I found that when using eval mode, the network's parformance is really bad.

Does this happen when you're using the non-efficient DenseNet as well? My hunch is that this may be a problem with your model choice, rather than an issue with this repo's implementation.

**Comment by gpleiss** [comment]
Also, not to be picky, but I don't think this PR is the right venue for these questions 😀 
If you're having an issue with this code, it's best to open up a new issue.
